exp/addons/internal/controllers/clusterresourcesetbinding_controller.go

```diff
@@ -79,33 +79,35 @@ func (r *ClusterResourceSetBindingReconciler) Reconcile(ctx context.Context, req
 		// Error reading the object - requeue the request.
 		return ctrl.Result{}, err
 	}
+	patchHelper, err := patch.NewHelper(binding, r.Client)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to configure the patch helper")
+	}
+	defer func() {
+		if err = patchHelper.Patch(ctx, binding); err != nil {
+			reterr = errors.Wrap(err, "failed to patch ClusterResourceSetBinding to add spec.clusterName")
+		}
+	}()
+
 	// Before 1.4 cluster name was stored as ownerRef; this code migrated this value to
 	// the spec.clusterName field if necessary.
-	clusterName := binding.Spec.ClusterName
-	if clusterName == "" {
-		// Initialize the patch helper.
-		patchHelper, err := patch.NewHelper(binding, r.Client)
+	if binding.Spec.ClusterName == "" {
+		clusterName, err := getClusterNameFromOwnerRef(binding.ObjectMeta)
 		if err != nil {
-			return ctrl.Result{}, errors.Wrap(err, "failed to configure the patch helper")
-		}
-		// Update the clusterName field of the existing ClusterResourceSetBindings with ownerReferences.
-		// More details please refer to: https://github.com/kubernetes-sigs/cluster-api/issues/7669.
-		if clusterName, err = getClusterNameFromOwnerRef(binding.ObjectMeta); err != nil {
 			return ctrl.Result{}, err
 		}
 		binding.Spec.ClusterName = clusterName
-		binding.OwnerReferences = util.RemoveOwnerRef(binding.OwnerReferences, metav1.OwnerReference{
-			APIVersion: clusterv1.GroupVersion.String(),
-			Kind:       "Cluster",
-			Name:       clusterName,
-		})
-		if err = patchHelper.Patch(ctx, binding); err != nil {
-			return ctrl.Result{}, errors.Wrap(err, "failed to patch ClusterResourceSetBinding to add spec.clusterName")
-		}
-		// Will enter the next Reconcile
+		// Return here to ensure this change is patched before proceeding.
 		return ctrl.Result{}, nil
 	}
-	cluster, err := util.GetClusterByName(ctx, r.Client, req.Namespace, clusterName)
+	// Remove the Cluster ownerRef if one is set.
+	binding.SetOwnerReferences(util.RemoveOwnerRef(binding.OwnerReferences, metav1.OwnerReference{
```

Comment:
It would be nice to a have a follow-up issue to try to figure out when/if we can get rid of this migration code
(one question is when, the other one is which code exactly can we get rid of after x releases and which do we have to keep forever)

(of course potentially also includes other places like the KCP controller)